### PR TITLE
chore: oracle tutorial for Vote Extensions

### PR DESCRIPTION
**Pull request by samricotta**
This is another tutorial for vote extensions but more advanced as we expect them to have their own chain etc. The snippets is based off the oracle repo https://github.com/facundomedica/oracle/tree/main and similar to the mitigation of auction front-running tutorial, this repo will be moved to `sdk-tutorials-repo`

Closes: https://github.com/cosmos/cosmos-sdk-docs/issues/92

### CODE REVIEW

**Comment by tac0turtle** [comment]
why does this live here instead of the tutorial repo?

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+* `ExtendVote` to get information from external price APIs.
+* `VerifyVoteExtension` to check that the format of the provided votes is correct.
+* `PrepareProposal` to process the vote extensions from the previous block and include them into the proposal as a transaction.
+* `ProcessProposal` to check that the first transaction in the proposal is actually a “special tx” that contains the price information.
+* `PreBlocker` to make price information available during FinalizeBlock.
+
+
+## Implement ExtendVote
+
+First we’ll create the `OracleVoteExtension` struct, this is the object that will be marshaled as bytes and signed by the validator.
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:24]: where am i suppose to create this? are there prerequisites to starting this tutorial? repos to clone, files to create ?

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+In our example we’ll use JSON to marshal the vote extension for simplicity but we recommend to find an encoding that produces a smaller output, given that large vote extensions could impact CometBFT’s performance. Custom encodings and compressed bytes can be used out of the box.
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:26]:
```suggestion
In our example we’ll use [JSON](might as well link to json) to marshal the vote extension for simplicity but we recommend to find an encoding that produces a smaller output, given that large vote extensions could impact CometBFT’s performance. Custom encodings and compressed bytes can be used out of the box.
```

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+	logger          log.Logger
+	currentBlock    int64                            // current block height
+	lastPriceSyncTS time.Time                        // last time we synced prices
+	providerTimeout time.Duration                    // timeout for fetching prices from providers
+	providers       map[string]Provider              // mapping of provider name to provider (e.g. Binance -> BinanceProvider)
+	providerPairs   map[string][]keeper.CurrencyPair // mapping of provider name to supported pairs (e.g. Binance -> [ATOM/USD])
+
+	Keeper keeper.Keeper // keeper of our oracle module
+}
+```
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:49]: would be good to touch on why some of the fields are needed

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+Finally, a function that returns `sdk.ExtendVoteHandler` is needed too, and this is where our vote extension logic will live.
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:51]:
```suggestion
Finally, a function that returns `sdk.ExtendVoteHandler` is needed, and this is where our vote extension logic will live.
```

maybe a short explanation on what the function does? the tutorial is assuming people need to read the code in order to understand how the oracle works

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        if err := h.verifyOraclePrices(ctx, voteExt.Prices); err != nil {
+            return nil, fmt.Errorf("failed to verify oracle prices from validator %X: %w", req.ValidatorAddress, err)
+        }
+
+        return &abci.ResponseVerifyVoteExtension{Status: abci.ResponseVerifyVoteExtension_ACCEPT}, nil
+    }
+}
+```
+
+## Implement PrepareProposal
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:108]: why do we need prepare proposal?

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+func (h *ProposalHandler) PrepareProposal() sdk.PrepareProposalHandler {
+    return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+        err := baseapp.ValidateVoteExtensions(ctx, h.valStore, req.Height, ctx.ChainID(), req.LocalLastCommit)
+        if err != nil {
+            return nil, err
+        }
+...
+```
+
+Then we proceed to make the calculations only if the current height if higher than the height at which vote extensions have been enabled. Remember that vote extensions are made available to the block proposer on the next block at which they are produced/enabled.
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:140]:
```suggestion
Then we proceed to make the calculations only if the current height is higher than the height at which vote extensions have been enabled. Remember that vote extensions are made available to the block proposer on the next block at which they are produced/enabled.
```

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+            }
+
+            injectedVoteExtTx := StakeWeightedPrices{
+                StakeWeightedPrices: stakeWeightedPrices,
+                ExtendedCommitInfo:  req.LocalLastCommit,
+            }
+...
+```
+
+Finally we inject the result as a transaction at a specific location, usually at the beginning of the block:
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:159]:
```suggestion
Finally, we inject the result as a transaction at a specific location, usually at the beginning of the block:
```

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+## Implement ProcessProposal
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:161]: why do we need process proposal

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+            var injectedVoteExtTx StakeWeightedPrices
+            if err := json.Unmarshal(req.Txs[0], &injectedVoteExtTx); err != nil {
+                h.logger.Error("failed to decode injected vote extension tx", "err", err)
+                return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, nil
+            }
+...
+```
+
+Then we re-validate the vote extensions signatures using
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:179]:
```suggestion
Next, we re-validate the vote extensions signatures using
```
nit

File: docs/user/tutorials/vote-extensions/oracle.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+}
+```
+
+Important: In this example we avoided using the mempool and other basics, please refer to the DefaultProposalHandler for a complete implementation: [https://github.com/cosmos/cosmos-sdk/blob/v0.50.1/baseapp/abci_utils.go](https://github.com/cosmos/cosmos-sdk/blob/v0.50.1/baseapp/abci_utils.go)
+
+## Implement PreBlocker
+
+Now validators are extending their vote, verifying other votes and including the result in the block. But how do we actually make use of this result? This is done in the PreBlocker which is code that is run before any other code during FinalizeBlock so we make sure we make this information available to the chain and its modules during the entire block execution (from BeginBlock).
+
+At this step we know that the injected tx is well-formatted and has been verified by the validators participating in consensus, so making use of it is straightforward. Just check if vote extensions are enabled, pick up the first transaction and use a method in your module’s keeper to set the result.
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/oracle.md:212]:
```suggestion
At this step we know that the injected transaction is well-formatted and has been verified by the validators participating in consensus, making use of it is straightforward. Check if vote extensions are enabled, pick up the first transaction and use a method in your module’s keeper to set the result.
```

File: docs/user/tutorials/vote-extensions/oracle.md (hunk 7 above)
Comment (samricotta) [reply docs/user/tutorials/vote-extensions/oracle.md:159]: this is true

File: docs/user/tutorials/vote-extensions/oracle.md (hunk 1 above)
Comment (facundomedica) [reply docs/user/tutorials/vote-extensions/oracle.md:24]: the pre-requisites are:

```
In this tutorial, we expect the reader to have a chain project already working as we won’t go through the steps of creating a new chain/module.

We also assume you are already familiar with the Cosmos SDK, if you are not we suggest you start with [https://tutorials.cosmos.network](https://tutorials.cosmos.network), as ABCI++ is considered an advanced topic.
```

**Review by facundomedica — APPROVED** [review]
LGTM, ty!!

**Pull request closed without merging**